### ISSUE TITLE
Broken decimal parsing when using BeforeValidator

### ISSUE THREAD

**Issue by OhioDschungel6** [pydantic/pydantic#9507]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

The [Pydantic docs ](https://docs.pydantic.dev/latest/concepts/conversion_table/#__tabbed_1_3) specify that decimal.Decimal can be properly parsed even when in strict mode.
This works fine, when not using a BeforeValidator. But when using a noop BeforeValidator this breaks, as the values are already parsed as floats.

I have added example code for better understanding:
`from decimal import Decimal
from typing import Any

from pydantic import BaseModel, model_validator

class M(BaseModel):
    a: Decimal
    b: Decimal
    c: Decimal

    @model_validator(mode="before")
    @classmethod
    def val(cls, value: Any) -> Any:
        return value

if __name__ == "__main__":
    json = """{"a": 0.1, "b": 0.1, "c": 0.1}"""
    m = M.model_validate_json(json, strict=True)`
When you delete the "val"-method everything works as expected.

Moreover decimal seems to be broken on another level.
`class M(BaseModel):
    a: Decimal

if __name__ == "__main__":
    json = """{"a": 9007199254740993.0}"""
    m = M.model_validate_json(json, strict=True)
    print(m.a == Decimal("9007199254740993")) #False
    print(m.a == Decimal(9007199254740993)) #False
    print(m.a == 9007199254740992.0) #True
    print(m.a == 9007199254740993.0) #True` 

This means I will lose precision even when using Decimals. 
The [doc](https://docs.pydantic.dev/2.0/usage/types/number_types/) specifies:

> Pydantic attempts to convert the value to a string, then passes the string to Decimal(v).

Therefore i would expect m.a == Decimal("9007199254740993") to be true

### Python, Pydantic & OS Version

```Text
pydantic version: 2.7.1
        pydantic-core version: 2.18.2
          pydantic-core build: profile=release pgo=true
               python version: 3.12.3 (tags/v3.12.3:f6650f9, Apr  9 2024, 14:05:25) [MSC v.1938 64 bit (AMD64)]
                     platform: Windows-10-10.0.19045-SP0
             related packages: mypy-1.10.0 typing_extensions-4.11.0
```

**Comment by Viicos** [comment]
This is the expected behavior of the _before_ (model) validator, the input is untouched. You probably want to use an _after_ model validator: in this case, `value["a"]` will be a `Decimal` instance. If you need more flexibility, you can also use a _wrap_ validator

**Comment by OhioDschungel6** [comment]
Please note, that I am not doing any validation. I am using the BeforeValidator as simple hook, to do something else before the validation starts. 
From my POV a noop BeforeValidator should not break the validation. 
From the [docs ](https://docs.pydantic.dev/2.3/usage/types/number_types/) i would expect that a, b and c are strings at the before validation stage.
Also decimal seems to have another problem with lossy data, which may or may not be related to the data being passed as float instead of a string.

**Comment by OhioDschungel6** [comment]
After thinking some more about it, i guess that pydantic is switching from json validation to python validation, when a BeforeValidator exists. This would be quite strange behaviour.

On the data type in the BeforeValidator side note: probably everything should be a string at this stage. For ints it is no problem, but floats have already lost precision at this point (I can not get it back at this stage), if i want to do my own validation.

**Comment by sydney-runkle** [comment]
Hmm, I would still expect this to work with a before validator. I'll chat with the team first thing tomorrow and solidify whether or not this is expected behavior.

Thanks for reporting this issue!

**Comment by Viicos** [comment]
Yes my bad it's only with a plain validator that nothing gets called after the plain validator!

**Comment by OhioDschungel6** [comment]
Oke perfect, thanks :)

**Comment by sydney-runkle** [comment]
Indeed, this is a bug!

**Comment by Viicos** [comment]
Progress for this will be tracked in https://github.com/pydantic/pydantic/issues/9507.
